strategies/QNetworkStrategy.py

```diff
@@ -0,0 +1,119 @@
+
+    def take_action(self, state):
+        if np.random.rand() < self.exploration_rate:
+            return random.randrange(0, self.action_space)
+        q_values = self.model.predict(state)
+        return np.argmax(q_values[0])
+
+    def experience_replay(self):
+        if len(self.memory) < BATCH_SIZE:
+            return
+        else:
+            minibatch = random.sample(self.memory, BATCH_SIZE)
+            for state, action, reward, state_next, done in minibatch:
+                Q = reward
+                if not done:
+                    Q = (reward + GAMMA * np.amax(self.model.predict(state_next)[0]))
+                Q_values = self.model.predict(state)
+                Q_values[0][action] = Q
+                self.model.fit(state, Q_values, verbose=0)
+            self.exploration_rate *= EXPLORATION_DECAY
+            self.exploration_rate = max(EXPLORATION_MIN, self.exploration_rate)
+
+    def get_model(self):
+        return self.model
+
+
+class TrainSolver:
+
+    def __init__(self, max_episodes):
+        self.max_episodes = max_episodes
+        self.score_table = deque(maxlen=400)
+        self.average_of_last_runs = None
+        self.model = None
+        self.play_episodes = 100
+        env = gym.make('CartPole-v1')
+        observation_space = env.observation_space.shape[0]
+        action_space = env.action_space.n
+        self.solver = Network(observation_space, action_space)
+
+    def train(self):
+        env = gym.make('CartPole-v1')
+        observation_space = env.observation_space.shape[0]
+        action_space = env.action_space.n
+
+        print("---------------------------------")
+        print("Solver starts")
+        print("---------------------------------")
+
+        self.model = self.solver.get_model()
+        episode = 0
+        while episode < self.max_episodes:
+
+            episode += 1
+            state = env.reset()
+            state = np.reshape(state, [1, observation_space])
+            step = 0
+            while True:
+
+                step += 1
+                # env.render()
+                action = self.solver.take_action(state)
+                state_next, reward, done, info = env.step(action)
+                if not done:
+                    reward = reward
+                else:
+                    reward = -reward
+                state_next = np.reshape(state_next, [1, observation_space])
+                self.solver.add_to_memory(state, action, reward, state_next, done)
+                state = state_next
+
+                if done:
+                    print("Run: " + str(episode) + ", exploration: " + str(self.solver.exploration_rate) + ", score: " + str(step))
+                    break
+                self.solver.experience_replay()
+
+    def return_trained_model(self):
+        return self.model
+
+    def save_model(self):
+        self.model.save('cartpole_model.h5')
+
+if __name__ == '__main__':
```

Comment:
Are you going to try to put this model into a Strategy?

Reply:
This will be done at a later time